dali/pipeline/operators/fused/crop_mirror_normalize.cu

```diff
@@ -21,104 +21,77 @@
+      // We won't change the underlying type after the first allocation
+      if (!kernel_sample_args_.has_value()) {
+        kernel_sample_args_ = std::vector<Args>(batch_size_);
       }
-    }
-
-    kernels::KernelRequirements req = kernel.Setup(ctx, in_view, per_sample_args);
-
-    output.set_type(TypeInfo::Create<OutputType>());
-    output.SetLayout(input_layout);
-    output.Resize(req.output_shapes[0]);
-
-    scratch_alloc.Reserve(req.scratch_sizes);
-    auto scratchpad = scratch_alloc.GetScratchpad();
-    ctx.scratchpad = &scratchpad;
-
-    auto out_view = view<OutputType, NumDims>(output);
-    kernel.Run(ctx, out_view, in_view, per_sample_args);
-  ),  // NOLINT
-  (
-    DALI_FAIL("Not supported number of dimensions: " + std::to_string(number_of_dims));
-  ));  // NOLINT
+      auto &kernel_sample_args = any_cast<std::vector<Args>&>(kernel_sample_args_);
+      kernel_sample_args[data_idx] = detail::GetKernelArgs<Dims, OutputType>(
+          input_layout_, output_layout_, slice_anchors_[data_idx], slice_shapes_[data_idx],
+          mirror_[data_idx], pad_output_, mean_vec_, inv_std_vec_);
+      // NOLINTNEXTLINE(whitespace/parens)
+    ), DALI_FAIL("Not supported number of dimensions: " + std::to_string(number_of_dims)););
+  );  // NOLINT(whitespace/parens)
 }
 
-}  // namespace detail
-
 template <>
-void CropMirrorNormalize<GPUBackend>::DataDependentSetup(DeviceWorkspace *ws) {
-  const auto &input = ws->Input<GPUBackend>(0);
-  for (int sample_idx = 0; sample_idx < batch_size_; sample_idx++) {
-    SetupSample(sample_idx, input_layout_, input.tensor_shape(sample_idx));
-    mirror_[sample_idx] = spec_.GetArgument<int>("mirror", ws, sample_idx);
-  }
-  auto &output = ws->Output<GPUBackend>(0);
-  output.SetLayout(output_layout_);
+bool CropMirrorNormalize<GPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                                const DeviceWorkspace &ws) {
+  output_desc.resize(1);
+  SetupAndInitialize(ws);
+  const auto &input = ws.Input<GPUBackend>(0);
+  auto &output = ws.OutputRef<GPUBackend>(0);
+  std::size_t number_of_dims = input.shape().sample_dim();
+  DALI_TYPE_SWITCH_WITH_FP16_GPU(input_type_, InputType,
+    DALI_TYPE_SWITCH_WITH_FP16_GPU(output_type_, OutputType,
+      VALUE_SWITCH(number_of_dims, Dims, (3, 4),
+      (
+        using Kernel = kernels::SliceFlipNormalizePermutePadGPU<OutputType, InputType, Dims>;
+        using Args = kernels::SliceFlipNormalizePermutePadArgs<Dims, OutputType>;
+        auto &kernel_sample_args = any_cast<std::vector<Args>&>(kernel_sample_args_);
+        output_desc[0].type = TypeInfo::Create<OutputType>();
+        output_desc[0].shape.resize(batch_size_, Dims);
+        kmgr_.Initialize<Kernel>();
+
+        kernels::KernelContext ctx;
+        ctx.gpu.stream = ws.stream();
+        auto in_view = view<const InputType, Dims>(input);
+        auto &req = kmgr_.Setup<Kernel>(0, ctx, in_view, kernel_sample_args);
```

Comment:
`const int kernel_instance = 0;`
for readability

Reply:
I would need to put it in the Run as well, and then we have the same 0 named twice in two different function. With the current approach that we do not create separate specializations for GPU and CPU I cannot define it as a static member variable, because it would be visible on CPU as well. Do you have better suggestion?

Reply:
I just meant to have a name constant here just for readability:
```
const int kernel_instance = 0;
auto &req = kmg_.Setup<Kernel>(kernel_instance, ...);
```
but feel free to ignore the suggestion if you think it doesn't make sense

Reply:
I mean, below we also have: `kmgr_.Run<Kernel>(0, 0, ctx, out_view, in_view, kernel_sample_args);`
It would make sense to introduce it on class level, but it would also mean that we have one kernel_instace on CPU. Not ideal. I'll just leave it as is.